qiskit/circuit/library/generalized_gates/unitary.py

```diff
@@ -161,6 +162,10 @@ def _define(self):
             )
 
             self.definition = qs_decomposition(self.to_matrix())
+            if not (
+                matrix_equal(Operator(self.definition).to_matrix(), self.to_matrix(), atol=1e-7)
+            ):
+                self.definition = Isometry(self.matrix, 0, 0).definition
```

Comment:
code updated to use Isometry we the calculated Operator is different than the original matrix.